data_safe_haven/infrastructure/programs/sre/firewall.py

```diff
@@ -0,0 +1,270 @@
+                            ],
+                            source_addresses=props.subnet_user_services_software_repositories_prefixes,
+                            target_fqdns=PermittedDomains.SOFTWARE_REPOSITORIES_PYTHON,
+                        ),
+                    ],
+                ),
+                network.AzureFirewallApplicationRuleCollectionArgs(
+                    action=network.AzureFirewallRCActionArgs(
+                        type=network.AzureFirewallRCActionType.ALLOW
+                    ),
+                    name="workspaces",
+                    priority=FirewallPriorities.SRE_WORKSPACES,
+                    rules=[
+                        network.AzureFirewallApplicationRuleArgs(
+                            description="Allow external Ubuntu keyserver requests",
+                            name="AllowUbuntuKeyserver",
+                            protocols=[
+                                network.AzureFirewallApplicationRuleProtocolArgs(
+                                    port=int(Ports.CLAMAV),
+                                    protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTP,
+                                ),
+                            ],
+                            source_addresses=props.subnet_workspaces_prefixes,
+                            target_fqdns=PermittedDomains.UBUNTU_KEYSERVER,
+                        ),
+                    ],
+                ),
+            ],
+            azure_firewall_name=f"{stack_name}-firewall",
+            ip_configurations=[
+                network.AzureFirewallIPConfigurationArgs(
+                    name="FirewallIpConfiguration",
+                    public_ip_address=network.SubResourceArgs(id=public_ip.id),
+                    subnet=network.SubResourceArgs(id=props.subnet_firewall_id),
+                )
+            ],
+            location=props.location,
+            management_ip_configuration=network.AzureFirewallIPConfigurationArgs(
+                name="FirewallManagementIpConfiguration",
+                public_ip_address=network.SubResourceArgs(id=public_ip_management.id),
+                subnet=network.SubResourceArgs(id=props.subnet_firewall_management_id),
+            ),
+            resource_group_name=props.resource_group_name,
+            sku=network.AzureFirewallSkuArgs(
+                name=network.AzureFirewallSkuName.AZF_W_V_NET,
+                tier=network.AzureFirewallSkuTier.BASIC,
+            ),
+            opts=child_opts,
+            tags=child_tags,
+        )
+
+        # Retrieve the private IP address for the firewall
+        private_ip_address = firewall.ip_configurations.apply(
+            lambda cfgs: (
+                "" if not cfgs else next(cfg.private_ip_address for cfg in cfgs)
+            )
+        )
+
+        # Route all connected traffic through the firewall
+        network.Route(
+            f"{self._name}_route_via_firewall",
+            address_prefix="0.0.0.0/0",
```

Comment:
Can we put this in an Enum?

Reply:
We don't have any enums of CIDRs and I don't think we even use explicit CIDRs anywhere else. I'll add a comment here in lieu of enum-ifying it.